bftengine/src/bftengine/ClientsManager.hpp

```diff
@@ -244,15 +249,22 @@ class ClientsManager : public ResPagesClient<ClientsManager>, public IPendingReq
 
   class RepliesInfo {
    public:
-    void deleteOldestReplyIfNeededSafe(NodeIdType clientId, uint16_t maxNumOfReqsPerClient);
-    bool insertOrAssignSafe(ReqId reqSeqNum, Time time);
+    // The thread safety here is between the PreProcessor thread, which is read-only, and the ReplicaImp main thread,
+    // which performs either read or write at a time. So, all write operations should be guarded as well as PrePropessor
+    // calls hasReply and isClientRequestInProcess. All other operations are safe.
+    void deleteReplyIfNeededSafe(NodeIdType clientId,
+                                 ReqId reqSeqNum,
+                                 uint16_t maxNumOfReqsPerClient,
+                                 uint16_t reqIndex);
+    void insertOrAssignSafe(ReqId reqSeqNum, uint16_t reqIndexInBatch);
     bool findSafe(ReqId reqSeqNum);
 
     bool find(ReqId reqSeqNum) const;
+    uint16_t getIndex(ReqId reqSeqNum) const;
 
    public:
     std::mutex repliesMapMutex_;
-    std::map<ReqId, Time> repliesMap_;
+    boost::bimaps::bimap<ReqId, uint16_t> repliesBiMap_;  // mapping request sequence number to request index in batch
```

Comment:
Right, it contains unique sequence numbers and unique indices.
The purpose of this structure here is to get the index given the seqNum and vice versa (to avoid map loop when we have the seqNum, as it was before).
Regarding the uniqueness - a sequence number will always have the same index because if a request was sent in a batch, the client will re-send it, if needed, as part of the same batch retransmit, so the request always keeps its index. (If the request was sent without batch, its index is 0 and will be the same when the request is retransmit.)
The indices are also unique within the repliesBiMap. This BiMap holds the currently saved replies in the reserved pages, so it's not possible to have two replies saved on the same page. deleteReplyIfNeededSafe() ensure this by deleting the overridden reply.